File: cat-core/src/main/java/com/dianping/cat/message/spi/internal/DefaultMessageTree.java
package com.dianping.cat.message.spi.internal;

import java.nio.charset.Charset;

import org.jboss.netty.buffer.ChannelBuffer;
import org.jboss.netty.buffer.ChannelBuffers;

import com.dianping.cat.message.Message;
import com.dianping.cat.message.spi.MessageTree;
import com.dianping.cat.message.spi.codec.PlainTextMessageCodec;

public class DefaultMessageTree implements MessageTree {
	private ChannelBuffer m_buf;

	private String m_domain;

	private String m_hostName;

	private String m_ipAddress;

	private Message m_message;

	private String m_messageId;

	private String m_parentMessageId;

	private String m_rootMessageId;

	private String m_sessionToken;

	private String m_threadGroupName;

	private String m_threadId;

	private String m_threadName;

	@Override
	public DefaultMessageTree copy() {
		DefaultMessageTree tree = new DefaultMessageTree();

		tree.setDomain(m_domain);
		tree.setHostName(m_hostName);
		tree.setIpAddress(m_ipAddress);
		tree.setMessageId(m_messageId);
		tree.setParentMessageId(m_parentMessageId);
		tree.setRootMessageId(m_rootMessageId);
		tree.setSessionToken(m_sessionToken);
		tree.setThreadGroupName(m_threadGroupName);
		tree.setThreadId(m_threadId);
		tree.setThreadName(m_threadName);
		tree.setMessage(m_message);

		return tree;
	}

	public ChannelBuffer getBuf() {
		return m_buf;
	}

	@Override
	public String getDomain() {
		return m_domain;
	}

	@Override
	public String getHostName() {
		return m_hostName;
	}

	@Override
	public String getIpAddress() {
		return m_ipAddress;
	}

	@Override
	public Message getMessage() {
		return m_message;
	}

	@Override
	public String getMessageId() {
		return m_messageId;
	}

	@Override
	public String getParentMessageId() {
		return m_parentMessageId;
	}

	@Override
	public String getRootMessageId() {
		return m_rootMessageId;
	}

	@Override
	public String getSessionToken() {
		return m_sessionToken;
	}

	public String getThreadGroupName() {
		return m_threadGroupName;
	}

	@Override
	public String getThreadId() {
		return m_threadId;
	}

	public String getThreadName() {
		return m_threadName;
	}

	public void setBuf(ChannelBuffer buf) {
		m_buf = buf;
	}

	@Override
	public void setDomain(String domain) {
		m_domain = domain;
	}

	@Override
	public void setHostName(String hostName) {
		m_hostName = hostName;
	}

	@Override
	public void setIpAddress(String ipAddress) {
		m_ipAddress = ipAddress;
	}

	@Override
	public void setMessage(Message message) {
		m_message = message;
	}

	@Override
	public void setMessageId(String messageId) {
		if (messageId != null && messageId.length() > 0) {
			m_messageId = messageId;
		}
	}

	@Override
	public void setParentMessageId(String parentMessageId) {
		if (parentMessageId != null && parentMessageId.length() > 0) {
			m_parentMessageId = parentMessageId;
		}
	}

	@Override
	public void setRootMessageId(String rootMessageId) {
		if (rootMessageId != null && rootMessageId.length() > 0) {
			m_rootMessageId = rootMessageId;
		}
	}

	@Override
	public void setSessionToken(String sessionToken) {
		m_sessionToken = sessionToken;
	}

	public void setThreadGroupName(String threadGroupName) {
		m_threadGroupName = threadGroupName;
	}

	@Override
	public void setThreadId(String threadId) {
		m_threadId = threadId;
	}

	public void setThreadName(String threadName) {
		m_threadName = threadName;
	}

	@Override
	public String toString() {
		PlainTextMessageCodec codec = new PlainTextMessageCodec();
		ChannelBuffer buf = ChannelBuffers.dynamicBuffer(8192);

		codec.encode(this, buf);
		buf.readInt(); // get rid of length
		return buf.toString(Charset.forName("utf-8"));
	}
}

File: lib/csharp/src/CatClient/Message/Internals/DefaultMessageTree.cs
using Org.Unidal.Cat.Message.Spi.Codec;
using System;
using Org.Unidal.Cat.Util;

namespace Org.Unidal.Cat.Message.Internals
{
    public class DefaultMessageTree : IMessageTree
    {
        private String _mMessageId;

        private String _mParentMessageId;

        private String _mRootMessageId;

        #region IMessageTree Members

        public IMessageTree Copy()
        {
            DefaultMessageTree tree = new DefaultMessageTree();

            tree.Domain = Domain;
            tree.HostName = HostName;
            tree.IpAddress = IpAddress;
            tree.MessageId = _mMessageId;
            tree.ParentMessageId = _mParentMessageId;
            tree.RootMessageId = _mRootMessageId;
            tree.SessionToken = SessionToken;
            tree.ThreadGroupName = ThreadGroupName;
            tree.ThreadId = ThreadId;
            tree.ThreadName = ThreadName;
            tree.Message = Message;
            tree.EstimatedByteSize = EstimatedByteSize;

            return tree;
        }

        public string Domain { get; set; }

        public string HostName { get { return NetworkInterfaceManager.HostName; } set { } } // set does nothing

        public string IpAddress { get { return NetworkInterfaceManager.HostIP; } set { } } // set does nothing

        public IMessage Message { get; set; }

        public String MessageId
        {
            get { return _mMessageId; }
            set
            {
                if (!string.IsNullOrEmpty(value))
                {
                    _mMessageId = value;
                }
            }
        }

        public String ParentMessageId
        {
            get { return _mParentMessageId; }
            set
            {
                if (!string.IsNullOrEmpty(value))
                {
                    _mParentMessageId = value;
                }
            }
        }

        public String RootMessageId
        {
            get { return _mRootMessageId; }
            set
            {
                if (!string.IsNullOrEmpty(value))
                {
                    _mRootMessageId = value;
                }
            }
        }

        public string SessionToken { get; set; }

        public string ThreadGroupName { get; set; }

        public string ThreadId { get; set; }

        public string ThreadName { get; set; }

        // Calculate byte size. If it is too big, take actions to prevent OOM.
        public int EstimatedByteSize { get; set;  }

        #endregion

        public override String ToString()
        {
            try
            {
                PlainTextMessageCodec codec = new PlainTextMessageCodec();
                ChannelBuffer buf = new ChannelBuffer(8192);

                codec.Encode(this, buf);

                buf.Reset();
                buf.Skip(4); // get rid of length

                return buf.ToString();
            }
            catch (Exception ex)
            {
                Cat.lastException = ex;
                return "";
            }
        }
    }
}